crates/worker/src/partition/snapshot_producer.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2023 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use anyhow::{anyhow, bail};
+use parking_lot::Mutex;
+use tokio::sync::oneshot::Sender;
+use tracing::{info, warn};
+
+use restate_core::TaskKind;
+use restate_partition_store::snapshots::{PartitionSnapshotMetadata, SnapshotFormatVersion};
+use restate_partition_store::PartitionStore;
+use restate_types::config::Configuration;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::live::LiveLoad;
+use restate_types::time::MillisSinceEpoch;
+
+pub struct SnapshotProducer {
+    pub partition_id: PartitionId,
+    pub config: Box<dyn LiveLoad<Configuration> + Send>,
+    latest_snapshot: Arc<Mutex<Option<SnapshotId>>>,
```

Comment:
From what I understood this only works as a guard to prevent multiple snapshot jobs to be running at the same time. Wondering if using a `tokio::sync::Semaphore` would be better here. 

A `SemaphorePermit` is also Send + Sync so you can send it safely across threads.

Reply:
If the Mutex is needed to also guard the value of the snapshot id I am wondering if using `tokio::sync::Mutex` to avoid locking the entire thread during the entire execution of the snapshot

Reply:
The check below seems to immediate reject creating new snapshots if one in progress. A simple and cheap way to do this is:
```
is_snapshot_in_progress: Arc<AtomicBool>,
```

Reply:
Switched to an atomic, thanks for the suggestions! :-)